web/client/reducers/queryform.js

```diff
@@ -0,0 +1,144 @@
+ * LICENSE file in the root directory of this source tree.
+ */
+
+const {ADD_FILTER_FIELD, REMOVE_FILTER_FIELD, UPDATE_FILTER_FIELD, UPDATE_EXCEPTION_FIELD} = require('../actions/queryform');
+const assign = require('object-assign');
+
+const initialState = {
+    filterFields: [
+        {
+            rowId: 0,
+            attribute: "",
+            operator: null,
+            value: null,
+            exception: null
+        }
+    ],
+    attributes: [
+        {
+            id: "Province",
+            type: "list",
+            values: [
+                "province1",
+                "province2",
+                "province3",
+                "province4",
+                "province5"
+            ]
+        },
+        {
+            id: "Municipality",
+            type: "list",
+            values: [
+                "municipality1",
+                "municipality2",
+                "municipality3",
+                "municipality4",
+                "municipality5"
+            ]
+        },
+        {
+            id: "RecoveryType",
+            type: "list",
+            values: [
+                "recoveryType1",
+                "recoveryType2",
+                "recoveryType3",
+                "recoveryType4",
+                "recoveryType5"
+            ]
+        },
+        {
+            id: "RejectType",
+            type: "list",
+            values: [
+                "rejectType1",
+                "rejectType2",
+                "rejectType3",
+                "rejectType4",
+                "rejectType5"
+            ]
+        },
+        {
+            id: "RecoveryOp",
+            type: "list",
+            values: [
+                "recoveryOp1",
+                "recoveryOp2",
+                "recoveryOp3",
+                "recoveryOp4",
+                "recoveryOp5"
+            ]
+        },
+        {
+            id: "ActionDate",
+            type: "date"
+        }
+    ]
+};
+
+function queryform(state = initialState, action) {
+    switch (action.type) {
+        case ADD_FILTER_FIELD: {
+            //
+            // Calculate the key number, this should be different for each new element
+            //
+            const newElement = {
+                rowId: new Date().getUTCMilliseconds(),
+                attribute: "",
+                operator: null,
+                value: null,
+                exception: null
+            };
+
+            return assign({}, state, {filterFields: [...state.filterFields, newElement]});
+        }
+        case REMOVE_FILTER_FIELD: {
+            return assign({}, state, {filterFields: state.filterFields.filter((field) => {
+                if (field.rowId !== action.rowId) {
+                    return field;
+                }
+            })});
+        }
+        case UPDATE_FILTER_FIELD: {
+            return assign({}, state, {filterFields: state.filterFields.map((field) => {
+                if (field.rowId === action.rowId) {
```

Comment:
if we name action.fieldName as the real name of the attribute to change, we can get rid of the switch and make it generic: field[action.fieldName] = action.fieldValue
Moreover you cannot assign new values to field, you have to create a new object, copying old values as usual assign({}, field, {[action.fieldName]: action.fieldValue})